astropy/io/ascii/core.py

```diff
@@ -1035,14 +1040,25 @@ def _validate_and_copy(col, converters):
         return converters_out
 
     def _convert_vals(self, cols):
-        for col in cols:
-            # If a specific dtype was specified for a column, then use that
-            # to set the defaults, otherwise use the generic defaults.
-            default_converters = ([convert_numpy(col.dtype)] if col.dtype
-                                  else self.default_converters)
-
-            # If the user supplied a specific convert then that takes precedence over defaults
-            converters = self.converters.get(col.name, default_converters)
+        # Loads of things happen before that change the number of columns.
+        # This is where we need the dtype that was passed in, so this is
+        # where we test.
+        if (self.dtype is not None) and len(self.dtype) != len(cols):
+            raise ValueError(f"Number of columns returned ({len(cols)}) does not match number of dtypes set ({len(self.dtype)}).")
+        for i, col in enumerate(cols):
+            # If the user set a dtype argument, that overrides everything else
+            if self.dtype is not None:
+                converters = [convert_numpy(self.dtype[i])]
+
+            # col.dtype is no longer used and is formally deprecated.
+            elif hasattr(col, '_dtype') and col._dtype is not None:
```

Comment:
This part is not the same logic as was here before, where `converters` can override anything inferred based on (the old) `col.dtype`. I think this should be:
```
for key, converters in self.converters.items():
    if fnmatch.fnmatch(col.name, key):
        break
else:
    # TODO: remove this check on a deprecated attribute in 5.1
    if getattr(col, '_dtype', None) is not None:
        converters = [convert_numpy(col.dtype)]
    else:
        converters = self.default_converters
```